# Peer-graded Assignment: Capstone Project Notebook
### Created by: Eric J. Puttock

# Assignment Week 2:

In this assignment, you will be required to explore, segment, and cluster the neighborhoods in the city of Toronto. However, unlike New York, the neighborhood data is not readily available on the internet. What is interesting about the field of data science is that each project can be challenging in its unique way, so you need to learn to be agile and refine the skill to learn new libraries and tools quickly depending on the project.

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. You will be required to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas dataframe so that it is in a structured format like the New York dataset.

Once the data is in a structured format, you can replicate the analysis that we did to the New York City dataset to explore and cluster the neighborhoods in the city of Toronto.

### Import all the libraries to be used for later just in case.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## Part 1:

We will read in data from the table from the following Wikipedia page listing postal codes of Canada.

In [2]:
WikiLink = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

The data that from the table is read into a pandas dataframe.

In [3]:
import pandas as pd
wikidf = pd.read_html(WikiLink)[0]
wikidf

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

We remove the rows of 'Borough' with 'Not assigned' from the dataframe (and reset index).

In [4]:
wikidf = wikidf[wikidf['Borough'] != 'Not assigned'].reset_index(drop=True)
wikidf

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Vic

If a cell has a 'Borough' but a 'Not assigned' 'Neighborhood', then the 'Neighborhood' should be assigned the same as borough.
However, as shown below, there aren't any 'Neighbourhood' rows with 'Not assigned' after removal of 'Not assigned' 'Borough.

Therefore, no additional effort will be made to code that portion.

In [5]:
wikidf[wikidf['Neighbourhood']=='Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

Here, we show our cleaned up dataframe contain 103 rows (Postal Codes) from Canada.

In [6]:
wikidf.shape

(103, 3)

## Part 2:

Remark: 
Here is my attempt at creating the Geospatial data dataframe (as instructed by the assignment using a while loop). I added a maximum iteration run and print outs so I can keep track and end the iteration. The following cell requires geocoder using Google API. However, Google-Geocode denies the request.

We are kindly provided with the Geospatial data in case we run into such problems for this assignment.

Please skip to "Import Geospatial data of Toronto".

In [7]:
# Install geocoder if needed:
# !conda install -c conda-forge geocoder --yes

In [8]:
import geocoder # import geocoder
# Attempt to find geolocatin of 'M1B'
postal_code = 'M1B'

# initialize your variable to None
lat_lng_coords = None
kk = 0
maxKK = 10

# loop until you get the coordinates (or max number of iterations).
while(lat_lng_coords is None and kk < maxKK):
    kk +=1
    query = '{}, Toronto, Ontario'.format(postal_code)
    g = geocoder.google(query)
    print(kk, g)
    lat_lng_coords = g.latlng

# Try to get coordinates. If it's empty, the exception would print out 'Did not get coordinates'.
try:
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    print(lat_lng_coords)
except:
    print('Did not get coordinates.')

1 <[REQUEST_DENIED] Google - Geocode [empty]>
2 <[REQUEST_DENIED] Google - Geocode [empty]>
3 <[REQUEST_DENIED] Google - Geocode [empty]>
4 <[REQUEST_DENIED] Google - Geocode [empty]>
5 <[REQUEST_DENIED] Google - Geocode [empty]>
6 <[REQUEST_DENIED] Google - Geocode [empty]>
7 <[REQUEST_DENIED] Google - Geocode [empty]>
8 <[REQUEST_DENIED] Google - Geocode [empty]>
9 <[REQUEST_DENIED] Google - Geocode [empty]>
10 <[REQUEST_DENIED] Google - Geocode [empty]>
Did not get coordinates.


### Import Geospatial data of Toronto.
Since, the above geocoder with google failed, I will continue with the provided CSV data as requested.

In [9]:
GS_url = 'http://cocl.us/Geospatial_data'

In [10]:
geodf = pd.read_csv(GS_url)
geodf.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We will join latitude & longitude data using Postal Code onto wikidf dataframe.

In [11]:
wikigeodf = wikidf.join(geodf.set_index('Postal Code'), on='Postal Code')
wikigeodf

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

## Part 3:

We will use geocoder to obtain the geographical coordinate of Toronto.
Install geocoder if needed, or geocoder was not used in Part 2.

In [12]:
# !conda install -c conda-forge geocoder --yes

In [13]:
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinate of Toronto is {}, {}.'.format(latitude, longitude))

The geographical coordinate of Toronto is 43.6534817, -79.3839347.


Now, we will take the map of Toronto and place markers in each Neighbourhood.

In [14]:
import folium

# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(wikigeodf['Latitude'], wikigeodf['Longitude'], wikigeodf['Borough'], wikigeodf['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)  
    
map_Toronto

In [15]:
print(wikigeodf.shape)
wikigeodf.head()

(103, 5)


Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [16]:
print(wikigeodf['Borough'].unique())

print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(wikigeodf['Borough'].unique()),
        wikigeodf.shape[0])
)

['North York' 'Downtown Toronto' 'Etobicoke' 'Scarborough' 'East York'
 'York' 'East Toronto' 'West Toronto' 'Central Toronto' 'Mississauga']
The dataframe has 10 boroughs and 103 neighborhoods.


Let's look into one of the Boroughs of Toronto: North York.

In [17]:
NorthYork_data = wikigeodf[wikigeodf['Borough'] == 'North York'].reset_index(drop=True)
NorthYork_data.head()

Postal Code     Borough                     Neighbourhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.464763  
3 -79.352188  
4 -79.445073

In [18]:
NorthYork_data.shape

(24, 5)

In [19]:
address = 'North York, Toronto'
geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of North York is {}, {}.'.format(latitude, longitude))

The geograpical coordinate of North York is 43.7543263, -79.44911696639593.


In [20]:
# create map of Manhattan using latitude and longitude values
map_NorthYork = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip( NorthYork_data['Latitude'], NorthYork_data['Longitude'], NorthYork_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_NorthYork)  
    
map_NorthYork

Here is a sample of data extraction of the first row from NorthYork_data.

In [21]:
neighbourhood_latitude = NorthYork_data.loc[12, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = NorthYork_data.loc[12, 'Longitude'] # neighborhood longitude value
neighbourhood_name = NorthYork_data.loc[12, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of York Mills, Silver Hills are 43.7574902, -79.37471409999999.


Enter Foursquare Credentials.
My credentials are hidden for obvious reasons!

In [22]:
## Hide:
#<--
CLIENT_ID = '-' # your Foursquare ID
CLIENT_SECRET = '-' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

#print('Your credentials:')
#print('CLIENT_ID: ' + CLIENT_ID)
#print('CLIENT_SECRET:' + CLIENT_SECRET)
#-->

Create getNearbyVenues to obtain venues from North York.

Send the get request to FourSquare API and examine the results for Nearby Venues.

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        if len(results)==0:
            print(r'**WARNING: {} does not have any categories within {} radius.**'.format(name,radius))
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [24]:
NorthYork_venues = getNearbyVenues(names=NorthYork_data['Neighbourhood'],
                                   latitudes=NorthYork_data['Latitude'],
                                   longitudes=NorthYork_data['Longitude']
                                  )

Parkwoods
Victoria Village
Lawrence Manor, Lawrence Heights
Don Mills
Glencairn
Don Mills
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Fairview, Henry Farm, Oriole
Northwood Park, York University
Bayview Village
Downsview
York Mills, Silver Hills
**WARNING: York Mills, Silver Hills does not have any categories within 500 radius.**
Downsview
North Park, Maple Leaf Park, Upwood Park
Humber Summit
Willowdale, Newtonbrook
Downsview
Bedford Park, Lawrence Manor East
Humberlea, Emery
Willowdale, Willowdale East
Downsview
York Mills West
Willowdale, Willowdale West


Let's check out the some of the venues (and their locations)

Remark: After reviewing the data, there are no recommendations for 'York Mills, Silver Hills' within radius of 500.

In [25]:
NorthYork_venues[NorthYork_venues['Neighbourhood']=='York Mills, Silver Hills']

Empty DataFrame
Columns: [Neighbourhood, Neighbourhood Latitude, Neighbourhood Longitude, Venue, Venue Latitude, Venue Longitude, Venue Category]
Index: []

Example: 'York Mills West' showing two categories within radius of 500.

In [26]:
NorthYork_venues[NorthYork_venues['Neighbourhood']=='York Mills West']

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
236  York Mills West               43.752758               -79.400049   
237  York Mills West               43.752758               -79.400049   

                 Venue  Venue Latitude  Venue Longitude     Venue Category  
236  Kitchen Food Fair       43.751298       -79.401393  Convenience Store  
237    Tournament Park       43.751257       -79.399717               Park

Let's see how many unique categories are in each neighbourhood.

Remark 1: We have duplicated Neighbourhoods within 'North York'. Example: Downsview is duplicated, but have different area codes. This presents a problem in terms of simply grouping and counting. This needs additional work to separate out correctly by area codes, or even the coodrinates themselves with Neighborhoods. See the second set dataframe table.

Remark 2: We lost at exactly one neighbourhood 'York Mills, Silver Hills'. There were no categories recommended within radius of 500.

In [27]:
NYvCounts = NorthYork_venues[['Neighbourhood','Venue Category']].groupby('Neighbourhood').count()
print("Originally, we considered {} neighborhoods of 'North York' borough.".format(NorthYork_data.shape[0]))
print('We only have {} rows neighborhoods.'.format(NYvCounts.shape[0]))
print('This is due to duplications of same named neighborhoods and some neighborhoods having no categories.')
NYvCounts

Originally, we considered 24 neighborhoods of 'North York' borough.
We only have 19 rows neighborhoods.
This is due to duplications of same named neighborhoods and some neighborhoods having no categories.


Venue Category
Neighbourhood                                                  
Bathurst Manor, Wilson Heights, Downsview North              23
Bayview Village                                               4
Bedford Park, Lawrence Manor East                            27
Don Mills                                                    28
Downsview                                                    13
Fairview, Henry Farm, Oriole                                 64
Glencairn                                                     4
Hillcrest Village                                             5
Humber Summit                                                 2
Humberlea, Emery                                              2
Lawrence Manor, Lawrence Heights                             10
North Park, Maple Leaf Park, Upwood Park                      5
Northwood Park, York University                               5
Parkwoods                                                     3
Victoria Village                                              6
Willowdale, Newtonbrook                                       1
Willowdale, Willowdale East                                  34
Willowdale, Willowdale West                                   4
York Mills West                                               2

In [28]:
NYvCounts2 = NorthYork_venues[['Neighbourhood','Neighbourhood Longitude','Venue Category']].groupby(['Neighbourhood','Neighbourhood Longitude']).count()
print("Originally, we considered {} neighborhoods of 'North York' borough.".format(NorthYork_data.shape[0]))
print('There are only {} neighborhoods that have at least one category for recommendation.'.format(NYvCounts2.shape[0]))
print('We have lost {} neighborhoods.'.format(NorthYork_data.shape[0]-NYvCounts2.shape[0]))
NYvCounts2

Originally, we considered 24 neighborhoods of 'North York' borough.
There are only 23 neighborhoods that have at least one category for recommendation.
We have lost 1 neighborhoods.


Venue Category
Neighbourhood                                   Neighbourhood Longitude                
Bathurst Manor, Wilson Heights, Downsview North -79.442259                           23
Bayview Village                                 -79.385975                            4
Bedford Park, Lawrence Manor East               -79.419750                           27
Don Mills                                       -79.352188                            6
                                                -79.340923                           22
Downsview                                       -79.520999                            4
                                                -79.506944                            4
                                                -79.495697                            2
                                                -79.464763                            3
Fairview, Henry Farm, Oriole                    -79.346556                           64
Glencairn                                       -79.445073                            4
Hillcrest Village                               -79.363452                            5
Humber Summit                                   -79.565963                            2
Humberlea, Emery                                -79.532242                            2
Lawrence Manor, Lawrence Heights                -79.464763                           10
North Park, Maple Leaf Park, Upwood Park        -79.490074                            5
Northwood Park, York University                 -79.487262                            5
Parkwoods                                       -79.329656                            3
Victoria Village                                -79.315572                            6
Willowdale, Newtonbrook                         -79.408493                            1
Willowdale, Willowdale East                     -79.408493                           34
Willowdale, Willowdale West                     -79.442259                            4
York Mills West                                 -79.400049                            2

In [29]:
print('There are {} unique categories.'.format(len(NorthYork_venues['Venue Category'].unique())),'within')

There are 106 unique categories. within


In [30]:
NorthYork_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude     Venue Category  
0         Brookbanks Park       43.751976       -79.332140               Park  
1             649 Variety       43.754513       -79.331942  Convenience Store  
2           Variety Store       43.751974       -79.333114  Food & Drink Shop  
3  Victoria Village Arena       43.723481       -79.315635       Hockey Arena  
4             Tim Hortons       43.725517       -79.313103        Coffee Shop

Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [31]:
# one hot encoding
NorthYork_onehot = pd.get_dummies(NorthYork_venues[['Venue Category']], prefix="", prefix_sep="")
# add neighborhood column back to dataframe
NorthYork_onehot['Neighbourhood'] = NorthYork_venues['Neighbourhood']
# We also add Longitude to prevent duplicated names from different area codes grouping together.
NorthYork_onehot['Neighbourhood Longitude'] = NorthYork_venues['Neighbourhood Longitude'] 
# move neighborhood column to the first column
fixed_columns = [NorthYork_onehot.columns[-1]] + list(NorthYork_onehot.columns[:-1])
NorthYork_onehot = NorthYork_onehot[fixed_columns]

# let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
#NorthYork_grouped = NorthYork_onehot.groupby('Neighbourhood').mean().reset_index()
# let's group rows by neighborhood and longtigude by taking the mean of the frequency of occurrence of each category
NorthYork_grouped = NorthYork_onehot.groupby(['Neighbourhood','Neighbourhood Longitude']).mean().reset_index()
NorthYork_grouped.shape

(23, 108)

In [32]:
NorthYork_grouped.head()

Neighbourhood  Neighbourhood Longitude  \
0  Bathurst Manor, Wilson Heights, Downsview North               -79.442259   
1                                  Bayview Village               -79.385975   
2                Bedford Park, Lawrence Manor East               -79.419750   
3                                        Don Mills               -79.352188   
4                                        Don Mills               -79.340923   

   Accessories Store  Airport  American Restaurant  Art Gallery  \
0                0.0      0.0             0.000000     0.000000   
1                0.0      0.0             0.000000     0.000000   
2                0.0      0.0             0.037037     0.000000   
3                0.0      0.0             0.000000     0.000000   
4                0.0      0.0             0.000000     0.045455   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Bakery  \
0                  0.0          0.000000            0.000000     0.0   
1                  0.0          0.000000            0.000000     0.0   
2                  0.0          0.000000            0.000000     0.0   
3                  0.0          0.000000            0.166667     0.0   
4                  0.0          0.045455            0.000000     0.0   

       Bank  Bar  Baseball Field  Basketball Court  Beer Store  Bike Shop  \
0  0.086957  0.0             0.0               0.0    0.000000   0.000000   
1  0.250000  0.0             0.0               0.0    0.000000   0.000000   
2  0.000000  0.0             0.0               0.0    0.000000   0.000000   
3  0.000000  0.0             0.0               0.0    0.000000   0.000000   
4  0.000000  0.0             0.0               0.0    0.090909   0.045455   

   Bookstore  Boutique  Breakfast Spot  Bridal Shop  Bubble Tea Shop  \
0        0.0       0.0        0.000000     0.043478         0.000000   
1        0.0       0.0        0.000000     0.000000         0.000000   
2        0.0       0.0        0.037037     0.000000         0.000000   
3        0.0       0.0        0.000000     0.000000         0.000000   
4        0.0       0.0        0.000000     0.000000         0.045455   

   Burger Joint  Burrito Place  Bus Station  Bus Stop   Butcher      Café  \
0           0.0            0.0          0.0       0.0  0.000000  0.000000   
1           0.0            0.0          0.0       0.0  0.000000  0.250000   
2           0.0            0.0          0.0       0.0  0.037037  0.037037   
3           0.0            0.0          0.0       0.0  0.000000  0.166667   
4           0.0            0.0          0.0       0.0  0.000000  0.045455   

   Caribbean Restaurant  Chinese Restaurant  Chocolate Shop  Clothing Store  \
0              0.000000            0.000000             0.0        0.000000   
1              0.000000            0.250000             0.0        0.000000   
2              0.000000            0.000000             0.0        0.000000   
3              0.166667            0.000000             0.0        0.000000   
4              0.000000            0.045455             0.0        0.045455   

   Coffee Shop  Comfort Food Restaurant  Construction & Landscaping  \
0     0.086957                 0.000000                    0.000000   
1     0.000000                 0.000000                    0.000000   
2     0.074074                 0.037037                    0.000000   
3     0.000000                 0.000000                    0.166667   
4     0.090909                 0.000000                    0.000000   

   Convenience Store  Cosmetics Shop  Deli / Bodega  Department Store  \
0           0.043478             0.0       0.043478               0.0   
1           0.000000             0.0       0.000000               0.0   
2           0.000000             0.0       0.000000               0.0   
3           0.000000             0.0       0.000000               0.0   
4           0.000000             0.0       0.000000               0.0   

   Dessert Shop  Dim Sum R

Let's find the top 5 venues from each neighbourhood.

Note: Some neighbourhoods may not have 5 venues. If they don't have additional venues 'No More Venues' will be returned.

In [33]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    # Some neighboorhoods don't have enough venues. In this case, we want to return 'No More Venues'.
    LRCS = len(row_categories_sorted.index.values[row_categories_sorted!=0])
    
    if LRCS < num_top_venues:
        output = row_categories_sorted.index.values[0:LRCS]
        for k in range(LRCS,num_top_venues):
            output=np.insert(output,len(output),'No More Venues')
        return output
    else:
        output = row_categories_sorted.index.values[0:num_top_venues]
        return output

In [34]:
num_top_venues = 5
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
#columns = ['Neighbourhood']
columns = ['Neighbourhood','Neighbourhood Longitude']

for ind in np.arange(num_top_venues):
    #columns.append('Most Common Venue')
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = NorthYork_grouped['Neighbourhood']
neighbourhoods_venues_sorted['Neighbourhood Longitude'] = NorthYork_grouped['Neighbourhood Longitude']

for ind in np.arange(NorthYork_grouped.shape[0]):
    #neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(NorthYork_grouped.iloc[ind, :], num_top_venues)
    neighbourhoods_venues_sorted.iloc[ind, 2:] = return_most_common_venues(NorthYork_grouped.iloc[ind, :], num_top_venues)
    
neighbourhoods_venues_sorted.tail()

Neighbourhood  Neighbourhood Longitude  \
18             Victoria Village               -79.315572   
19      Willowdale, Newtonbrook               -79.408493   
20  Willowdale, Willowdale East               -79.408493   
21  Willowdale, Willowdale West               -79.442259   
22              York Mills West               -79.400049   

   1st Most Common Venue       2nd Most Common Venue 3rd Most Common Venue  \
18     French Restaurant  Financial or Legal Service           Coffee Shop   
19                  Park               Women's Store        No More Venues   
20      Ramen Restaurant                        Café            Restaurant   
21           Coffee Shop                    Pharmacy           Pizza Place   
22                  Park           Convenience Store         Women's Store   

   4th Most Common Venue 5th Most Common Venue  
18           Pizza Place          Hockey Arena  
19        No More Venues        No More Venues  
20           Pizza Place        Sandwich Place  
21                  Bank         Women's Store  
22        No More Venues        No More Venues

Remark: Since we joined them using on group, we have a few duplicates of the data. We need to drop data that wasn't joined correctly. For exampe, Don Mills with duplicated Neighbourhoods, have few incorect joints. We can drop rows with Longitudes that do not agree with Neighbourhood Longitude (like a foregin key to a database). I need to figure out how to connect them using multiple index keys in the future.

In [35]:
NorthYork_merged = NorthYork_data
NorthYork_merged = NorthYork_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on=['Neighbourhood'])
NYm = NorthYork_merged.reset_index(drop=True)
print(NYm.shape)
NYm

(38, 11)


Postal Code     Borough                                    Neighbourhood  \
0          M3A  North York                                        Parkwoods   
1          M4A  North York                                 Victoria Village   
2          M6A  North York                 Lawrence Manor, Lawrence Heights   
3          M3B  North York                                        Don Mills   
4          M3B  North York                                        Don Mills   
5          M6B  North York                                        Glencairn   
6          M3C  North York                                        Don Mills   
7          M3C  North York                                        Don Mills   
8          M2H  North York                                Hillcrest Village   
9          M3H  North York  Bathurst Manor, Wilson Heights, Downsview North   
10         M2J  North York                     Fairview, Henry Farm, Oriole   
11         M3J  North York                  Northwood Park, York University   
12         M2K  North York                                  Bayview Village   
13         M3K  North York                                        Downsview   
14         M3K  North York                                        Downsview   
15         M3K  North York                                        Downsview   
16         M3K  North York                                        Downsview   
17         M2L  North York                         York Mills, Silver Hills   
18         M3L  North York                                        Downsview   
19         M3L  North York                                        Downsview   
20         M3L  North York                                        Downsview   
21         M3L  North York                                        Downsview   
22         M6L  North York         North Park, Maple Leaf Park, Upwood Park   
23         M9L  North York                                    Humber Summit   
24         M2M  North York                          Willowdale, Newtonbrook   
25         M3M  North York                                        Downsview   
26         M3M  North York                                        Downsview   
27         M3M  North York                                        Downsview   
28         M3M  North York                                        Downsview   
29         M5M  North York                Bedford Park, Lawrence Manor East   
30         M9M  North York                                 Humberlea, Emery   
31         M2N  North York                      Willowdale, Willowdale East   
32         M3N  North York                                        Downsview   
33         M3N  North York                                        Downsview   
34         M3N  North York                                        Downsview   
35         M3N  North York                                        Downsview   
36         M2P  North York                                  York Mills West   
37         M2R  North York                      Willowdale, Willowdale West   

     Latitude  Longitude  Neighbourhood Longitude     1st Most Common Venue  \
0   43.753259 -79.329656               -79.329656                      Park   
1   43.725882 -79.315572               -79.315572         French Restaurant   
2   43.718518 -79.464763               -79.464763            Clothing Store   
3   43.745906 -79.352188               -79.352188      Caribbean Restaurant   
4   43.745906 -79.352188               -79.340923                       Gym   
5   43.709577 -79.445073               -79.445073                      Park   
6   43.725900 -79.340923               -79.352188      Caribbean Restaurant   
7   43.725900 -79.340923               -79.340923                       Gym   
8   43.803762 -79.363452               -79.363452  Mediterranean Restaurant   
9   43.754328 -79.442259               -79.442259               Coffee Shop   
10  43.778517 -79.346556               -79.346556   

In [36]:
NYMedit = NYm[['Longitude','Neighbourhood Longitude']]
NYMedit.head()

Longitude  Neighbourhood Longitude
0 -79.329656               -79.329656
1 -79.315572               -79.315572
2 -79.464763               -79.464763
3 -79.352188               -79.352188
4 -79.352188               -79.340923

Identify the rows where the two columns don't agree.

In [37]:
drop_list = []
keep_list = []
for rownum in range(0,NYMedit.shape[0]):
    if NYMedit['Longitude'][rownum] != NYMedit['Neighbourhood Longitude'][rownum]:
        # Just to keep the special cases of when we don't have any venues categories.
        if np.isnan(NYMedit['Neighbourhood Longitude'][rownum]):
                keep_list.append(rownum)
                continue
        drop_list.append(rownum)
    else:
        keep_list.append(rownum)

print('We keep the numbers in rows' , keep_list)

We keep the numbers in rows [0, 1, 2, 3, 5, 7, 8, 9, 10, 11, 12, 16, 17, 19, 22, 23, 24, 27, 29, 30, 31, 32, 36, 37]


In [38]:
NYm_fixed = NYm.iloc[keep_list,:].reset_index(drop=True)
NYm_fixed.drop(columns=['Neighbourhood Longitude'],inplace=True)

Here's our nice organized dataframe!

In [39]:
NYm_fixed

Postal Code     Borough                                    Neighbourhood  \
0          M3A  North York                                        Parkwoods   
1          M4A  North York                                 Victoria Village   
2          M6A  North York                 Lawrence Manor, Lawrence Heights   
3          M3B  North York                                        Don Mills   
4          M6B  North York                                        Glencairn   
5          M3C  North York                                        Don Mills   
6          M2H  North York                                Hillcrest Village   
7          M3H  North York  Bathurst Manor, Wilson Heights, Downsview North   
8          M2J  North York                     Fairview, Henry Farm, Oriole   
9          M3J  North York                  Northwood Park, York University   
10         M2K  North York                                  Bayview Village   
11         M3K  North York                                        Downsview   
12         M2L  North York                         York Mills, Silver Hills   
13         M3L  North York                                        Downsview   
14         M6L  North York         North Park, Maple Leaf Park, Upwood Park   
15         M9L  North York                                    Humber Summit   
16         M2M  North York                          Willowdale, Newtonbrook   
17         M3M  North York                                        Downsview   
18         M5M  North York                Bedford Park, Lawrence Manor East   
19         M9M  North York                                 Humberlea, Emery   
20         M2N  North York                      Willowdale, Willowdale East   
21         M3N  North York                                        Downsview   
22         M2P  North York                                  York Mills West   
23         M2R  North York                      Willowdale, Willowdale West   

     Latitude  Longitude     1st Most Common Venue  \
0   43.753259 -79.329656                      Park   
1   43.725882 -79.315572         French Restaurant   
2   43.718518 -79.464763            Clothing Store   
3   43.745906 -79.352188      Caribbean Restaurant   
4   43.709577 -79.445073                      Park   
5   43.725900 -79.340923                       Gym   
6   43.803762 -79.363452  Mediterranean Restaurant   
7   43.754328 -79.442259               Coffee Shop   
8   43.778517 -79.346556            Clothing Store   
9   43.767980 -79.487262      Caribbean Restaurant   
10  43.786947 -79.385975                      Café   
11  43.737473 -79.464763                   Airport   
12  43.757490 -79.374714                       NaN   
13  43.739015 -79.506944                      Park   
14  43.713756 -79.490074          Basketball Court   
15  43.756303 -79.565963               Pizza Place   
16  43.789053 -79.408493                      Park   
17  43.728496 -79.495697                Food Truck   
18  43.733283 -79.419750        Italian Restaurant   
19  43.724766 -79.532242              Food Service   
20  43.770120 -79.408493          Ramen Restaurant   
21  43.761631 -79.520999            Discount Store   
22  43.752758 -79.400049                      Park   
23  43.782736 -79.442259               Coffee Shop   

         2nd Most Common Venue       3rd Most Common Venue  \
0            Food & Drink Shop           Convenience Store   
1   Financial or Legal Service                 Coffee Shop   
2       Furniture / Home Store                    Boutique   
3                         Café  Construction & Landscaping   
4             Sushi Restaurant         Japanese Restaurant   
5                   Restaurant                  Beer Store   
6                  Golf Course        Fast Food Restaurant   
7                         Bank          Frozen Yogurt Shop   
8                  Coffee Shop        Fast Food Restaurant   
9                Metro Station              Massage Stud

In [40]:
NYm_fixed.loc[:,'1st Most Common Venue':] = NYm_fixed.loc[:,'1st Most Common Venue':].fillna('No More Venues')

Now we have successfully wrangled with North York Borough data to extract the most common venues (if any).
If there aren't any venues, or enough venues, those features were filled with 'No More Venues'.

In [41]:
NYm_fixed.head()

Postal Code     Borough                     Neighbourhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude 1st Most Common Venue       2nd Most Common Venue  \
0 -79.329656                  Park           Food & Drink Shop   
1 -79.315572     French Restaurant  Financial or Legal Service   
2 -79.464763        Clothing Store      Furniture / Home Store   
3 -79.352188  Caribbean Restaurant                        Café   
4 -79.445073                  Park            Sushi Restaurant   

        3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  
0           Convenience Store         Women's Store         No More Venues  
1                 Coffee Shop           Pizza Place           Hockey Arena  
2                    Boutique     Accessories Store  Vietnamese Restaurant  
3  Construction & Landscaping    Athletics & Sports                    Gym  
4         Japanese Restaurant                   Pub          Women's Store

Now, let's try to see if we can group these into clusters using their top 5 most common venues.

### Make the Training set. This will consist of the top 5 venues of each row.
### We will use KMeans unsupervised clustering to attempt to cluster.

In [42]:
DataToFit = NYm_fixed.loc[:,'1st Most Common Venue':'5th Most Common Venue']
print(DataToFit.shape)

# Position of Missing Values: DataToFit.iloc[[12],:]

(24, 5)


Train the Model using 6 (arbitrary) clusters.

In [43]:
NorthYork_grouped.head()

Neighbourhood  Neighbourhood Longitude  \
0  Bathurst Manor, Wilson Heights, Downsview North               -79.442259   
1                                  Bayview Village               -79.385975   
2                Bedford Park, Lawrence Manor East               -79.419750   
3                                        Don Mills               -79.352188   
4                                        Don Mills               -79.340923   

   Accessories Store  Airport  American Restaurant  Art Gallery  \
0                0.0      0.0             0.000000     0.000000   
1                0.0      0.0             0.000000     0.000000   
2                0.0      0.0             0.037037     0.000000   
3                0.0      0.0             0.000000     0.000000   
4                0.0      0.0             0.000000     0.045455   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Bakery  \
0                  0.0          0.000000            0.000000     0.0   
1                  0.0          0.000000            0.000000     0.0   
2                  0.0          0.000000            0.000000     0.0   
3                  0.0          0.000000            0.166667     0.0   
4                  0.0          0.045455            0.000000     0.0   

       Bank  Bar  Baseball Field  Basketball Court  Beer Store  Bike Shop  \
0  0.086957  0.0             0.0               0.0    0.000000   0.000000   
1  0.250000  0.0             0.0               0.0    0.000000   0.000000   
2  0.000000  0.0             0.0               0.0    0.000000   0.000000   
3  0.000000  0.0             0.0               0.0    0.000000   0.000000   
4  0.000000  0.0             0.0               0.0    0.090909   0.045455   

   Bookstore  Boutique  Breakfast Spot  Bridal Shop  Bubble Tea Shop  \
0        0.0       0.0        0.000000     0.043478         0.000000   
1        0.0       0.0        0.000000     0.000000         0.000000   
2        0.0       0.0        0.037037     0.000000         0.000000   
3        0.0       0.0        0.000000     0.000000         0.000000   
4        0.0       0.0        0.000000     0.000000         0.045455   

   Burger Joint  Burrito Place  Bus Station  Bus Stop   Butcher      Café  \
0           0.0            0.0          0.0       0.0  0.000000  0.000000   
1           0.0            0.0          0.0       0.0  0.000000  0.250000   
2           0.0            0.0          0.0       0.0  0.037037  0.037037   
3           0.0            0.0          0.0       0.0  0.000000  0.166667   
4           0.0            0.0          0.0       0.0  0.000000  0.045455   

   Caribbean Restaurant  Chinese Restaurant  Chocolate Shop  Clothing Store  \
0              0.000000            0.000000             0.0        0.000000   
1              0.000000            0.250000             0.0        0.000000   
2              0.000000            0.000000             0.0        0.000000   
3              0.166667            0.000000             0.0        0.000000   
4              0.000000            0.045455             0.0        0.045455   

   Coffee Shop  Comfort Food Restaurant  Construction & Landscaping  \
0     0.086957                 0.000000                    0.000000   
1     0.000000                 0.000000                    0.000000   
2     0.074074                 0.037037                    0.000000   
3     0.000000                 0.000000                    0.166667   
4     0.090909                 0.000000                    0.000000   

   Convenience Store  Cosmetics Shop  Deli / Bodega  Department Store  \
0           0.043478             0.0       0.043478               0.0   
1           0.000000             0.0       0.000000               0.0   
2           0.000000             0.0       0.000000               0.0   
3           0.000000             0.0       0.000000               0.0   
4           0.000000             0.0       0.000000               0.0   

   Dessert Shop  Dim Sum R

Drop the columns that we don't need for Kmeans. We're only interested in the scoring of each venue.

In [44]:
NorthYork_grouped_clustering = NorthYork_grouped.drop(columns = ['Neighbourhood', 'Neighbourhood Longitude'],axis=1)
NorthYork_grouped_clustering.shape

(23, 106)

Adding in a row of 0's to take place for the one missing neighborhood (and re-order dataframe appropriately) 'York Mills, Silver Hills'.

In [45]:
NorthYork_grouped_clustering = NorthYork_grouped_clustering.append(pd.Series(np.zeros(NorthYork_grouped_clustering.shape[1]), index=NorthYork_grouped_clustering.columns), ignore_index=True)
order = [1,2,3,4,5,6,7,8,9,10,11,-1,12,13,14,15,16,17,18,19,20,21,22]
NorthYork_grouped_clustering.iloc[order,:].reset_index(drop=True,inplace=True)
NorthYork_grouped_clustering.shape

(24, 106)

In [46]:
NorthYork_grouped_clustering.head()

Accessories Store  Airport  American Restaurant  Art Gallery  \
0                0.0      0.0             0.000000     0.000000   
1                0.0      0.0             0.000000     0.000000   
2                0.0      0.0             0.037037     0.000000   
3                0.0      0.0             0.000000     0.000000   
4                0.0      0.0             0.000000     0.045455   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Bakery  \
0                  0.0          0.000000            0.000000     0.0   
1                  0.0          0.000000            0.000000     0.0   
2                  0.0          0.000000            0.000000     0.0   
3                  0.0          0.000000            0.166667     0.0   
4                  0.0          0.045455            0.000000     0.0   

       Bank  Bar  Baseball Field  Basketball Court  Beer Store  Bike Shop  \
0  0.086957  0.0             0.0               0.0    0.000000   0.000000   
1  0.250000  0.0             0.0               0.0    0.000000   0.000000   
2  0.000000  0.0             0.0               0.0    0.000000   0.000000   
3  0.000000  0.0             0.0               0.0    0.000000   0.000000   
4  0.000000  0.0             0.0               0.0    0.090909   0.045455   

   Bookstore  Boutique  Breakfast Spot  Bridal Shop  Bubble Tea Shop  \
0        0.0       0.0        0.000000     0.043478         0.000000   
1        0.0       0.0        0.000000     0.000000         0.000000   
2        0.0       0.0        0.037037     0.000000         0.000000   
3        0.0       0.0        0.000000     0.000000         0.000000   
4        0.0       0.0        0.000000     0.000000         0.045455   

   Burger Joint  Burrito Place  Bus Station  Bus Stop   Butcher      Café  \
0           0.0            0.0          0.0       0.0  0.000000  0.000000   
1           0.0            0.0          0.0       0.0  0.000000  0.250000   
2           0.0            0.0          0.0       0.0  0.037037  0.037037   
3           0.0            0.0          0.0       0.0  0.000000  0.166667   
4           0.0            0.0          0.0       0.0  0.000000  0.045455   

   Caribbean Restaurant  Chinese Restaurant  Chocolate Shop  Clothing Store  \
0              0.000000            0.000000             0.0        0.000000   
1              0.000000            0.250000             0.0        0.000000   
2              0.000000            0.000000             0.0        0.000000   
3              0.166667            0.000000             0.0        0.000000   
4              0.000000            0.045455             0.0        0.045455   

   Coffee Shop  Comfort Food Restaurant  Construction & Landscaping  \
0     0.086957                 0.000000                    0.000000   
1     0.000000                 0.000000                    0.000000   
2     0.074074                 0.037037                    0.000000   
3     0.000000                 0.000000                    0.166667   
4     0.090909                 0.000000                    0.000000   

   Convenience Store  Cosmetics Shop  Deli / Bodega  Department Store  \
0           0.043478             0.0       0.043478               0.0   
1           0.000000             0.0       0.000000               0.0   
2           0.000000             0.0       0.000000               0.0   
3           0.000000             0.0       0.000000               0.0   
4           0.000000             0.0       0.000000               0.0   

   Dessert Shop  Dim Sum Restaurant     Diner  Discount Store  \
0           0.0            0.000000  0.043478        0.000000   
1           0.0            0.000000  0.000000        0.000000   
2           0.0            0.000000  0.000000        0.000000   
3           0.0            0.000000  0.000000        0.000000   
4           0.0            0.045455  0.000000        0.045455   

   Distribution Center  Dog Run  Electronics Store  Event Space  \
0        

Run the Kmeans Algorithm to classify their groups.

In [47]:
# Set number of clusters
kclusters = 6
DataToFit = NorthYork_grouped_clustering
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=4).fit(DataToFit)
# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 0, 1, 1, 1, 3, 5, 1])

In [48]:
NYm_fixed.head()

Postal Code     Borough                     Neighbourhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude 1st Most Common Venue       2nd Most Common Venue  \
0 -79.329656                  Park           Food & Drink Shop   
1 -79.315572     French Restaurant  Financial or Legal Service   
2 -79.464763        Clothing Store      Furniture / Home Store   
3 -79.352188  Caribbean Restaurant                        Café   
4 -79.445073                  Park            Sushi Restaurant   

        3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  
0           Convenience Store         Women's Store         No More Venues  
1                 Coffee Shop           Pizza Place           Hockey Arena  
2                    Boutique     Accessories Store  Vietnamese Restaurant  
3  Construction & Landscaping    Athletics & Sports                    Gym  
4         Japanese Restaurant                   Pub          Women's Store

Add clustering labels to the fixed data table for summarization.

In [49]:
# add clustering labels
NYm_fixed.insert(5, 'Cluster Labels', kmeans.labels_)
NYm_fixed.head()

Postal Code     Borough                     Neighbourhood   Latitude  \
0         M3A  North York                         Parkwoods  43.753259   
1         M4A  North York                  Victoria Village  43.725882   
2         M6A  North York  Lawrence Manor, Lawrence Heights  43.718518   
3         M3B  North York                         Don Mills  43.745906   
4         M6B  North York                         Glencairn  43.709577   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.329656               1                  Park   
1 -79.315572               1     French Restaurant   
2 -79.464763               1        Clothing Store   
3 -79.352188               0  Caribbean Restaurant   
4 -79.445073               1                  Park   

        2nd Most Common Venue       3rd Most Common Venue  \
0           Food & Drink Shop           Convenience Store   
1  Financial or Legal Service                 Coffee Shop   
2      Furniture / Home Store                    Boutique   
3                        Café  Construction & Landscaping   
4            Sushi Restaurant         Japanese Restaurant   

  4th Most Common Venue  5th Most Common Venue  
0         Women's Store         No More Venues  
1           Pizza Place           Hockey Arena  
2     Accessories Store  Vietnamese Restaurant  
3    Athletics & Sports                    Gym  
4                   Pub          Women's Store

Visualize clustering on the Map.

In [50]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(NYm_fixed['Latitude'], NYm_fixed['Longitude'], NYm_fixed['Neighbourhood'], NYm_fixed['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Let's take a look at the data tables of the clusters.

In [51]:
Group0 = NYm_fixed[NYm_fixed['Cluster Labels']==0]
Group1 = NYm_fixed[NYm_fixed['Cluster Labels']==1]
Group2 = NYm_fixed[NYm_fixed['Cluster Labels']==2]
Group3 = NYm_fixed[NYm_fixed['Cluster Labels']==3]
Group4 = NYm_fixed[NYm_fixed['Cluster Labels']==4]
Group5 = NYm_fixed[NYm_fixed['Cluster Labels']==5]

In [52]:
Group0

Postal Code     Borough             Neighbourhood   Latitude  Longitude  \
3          M3B  North York                 Don Mills  43.745906 -79.352188   
12         M2L  North York  York Mills, Silver Hills  43.757490 -79.374714   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
3                0  Caribbean Restaurant                  Café   
12               0        No More Venues        No More Venues   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
3   Construction & Landscaping    Athletics & Sports                   Gym  
12              No More Venues        No More Venues        No More Venues

In [53]:
Group1

Postal Code     Borough                             Neighbourhood  \
0          M3A  North York                                 Parkwoods   
1          M4A  North York                          Victoria Village   
2          M6A  North York          Lawrence Manor, Lawrence Heights   
4          M6B  North York                                 Glencairn   
5          M3C  North York                                 Don Mills   
6          M2H  North York                         Hillcrest Village   
9          M3J  North York           Northwood Park, York University   
10         M2K  North York                           Bayview Village   
11         M3K  North York                                 Downsview   
14         M6L  North York  North Park, Maple Leaf Park, Upwood Park   
15         M9L  North York                             Humber Summit   
16         M2M  North York                   Willowdale, Newtonbrook   
18         M5M  North York         Bedford Park, Lawrence Manor East   
20         M2N  North York               Willowdale, Willowdale East   
21         M3N  North York                                 Downsview   
23         M2R  North York               Willowdale, Willowdale West   

     Latitude  Longitude  Cluster Labels     1st Most Common Venue  \
0   43.753259 -79.329656               1                      Park   
1   43.725882 -79.315572               1         French Restaurant   
2   43.718518 -79.464763               1            Clothing Store   
4   43.709577 -79.445073               1                      Park   
5   43.725900 -79.340923               1                       Gym   
6   43.803762 -79.363452               1  Mediterranean Restaurant   
9   43.767980 -79.487262               1      Caribbean Restaurant   
10  43.786947 -79.385975               1                      Café   
11  43.737473 -79.464763               1                   Airport   
14  43.713756 -79.490074               1          Basketball Court   
15  43.756303 -79.565963               1               Pizza Place   
16  43.789053 -79.408493               1                      Park   
18  43.733283 -79.419750               1        Italian Restaurant   
20  43.770120 -79.408493               1          Ramen Restaurant   
21  43.761631 -79.520999               1            Discount Store   
23  43.782736 -79.442259               1               Coffee Shop   

         2nd Most Common Venue 3rd Most Common Venue  \
0            Food & Drink Shop     Convenience Store   
1   Financial or Legal Service           Coffee Shop   
2       Furniture / Home Store              Boutique   
4             Sushi Restaurant   Japanese Restaurant   
5                   Restaurant            Beer Store   
6                  Golf Course  Fast Food Restaurant   
9                Metro Station        Massage Studio   
10                        Bank   Japanese Restaurant   
11                        Park              Bus Stop   
14                       Trail                  Park   
15                         Gym         Women's Store   
16               Women's Store        No More Venues   
18                  Restaurant           Coffee Shop   
20                        Café            Restaurant   
21               Grocery Store    Athletics & Sports   
23                    Pharmacy           Pizza Place   

         4th Most Common Venue  5th Most Common Venue  
0                Women's Store         No More Venues  
1                  Pizza Place           Hockey Arena  
2            Accessories Store  Vietnamese Restaurant  
4                          Pub          Women's Store  
5                  Coffee Shop              Bike Shop  
6                      Dog Run                   Pool  
9                  Coffee Shop                    Bar  
10          Chinese Restaurant          Women's Store  
11               Women's Store         No More Venues  
14  Construction & Landscaping                 Bakery  
15              No

In [54]:
Group2

Postal Code     Borough     Neighbourhood   Latitude  Longitude  \
19         M9M  North York  Humberlea, Emery  43.724766 -79.532242   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19               2          Food Service        Baseball Field   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
19         Women's Store        No More Venues        No More Venues

In [55]:
Group3

Postal Code     Borough                                    Neighbourhood  \
7          M3H  North York  Bathurst Manor, Wilson Heights, Downsview North   
13         M3L  North York                                        Downsview   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
7   43.754328 -79.442259               3           Coffee Shop   
13  43.739015 -79.506944               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7                   Bank    Frozen Yogurt Shop        Sandwich Place   
13         Grocery Store                  Bank         Shopping Mall   

   5th Most Common Venue  
7         Ice Cream Shop  
13         Women's Store

In [56]:
Group4

Postal Code     Borough    Neighbourhood   Latitude  Longitude  \
17         M3M  North York        Downsview  43.728496 -79.495697   
22         M2P  North York  York Mills West  43.752758 -79.400049   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
17               4            Food Truck        Baseball Field   
22               4                  Park     Convenience Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
17         Women's Store        No More Venues        No More Venues  
22         Women's Store        No More Venues        No More Venues

In [57]:
Group5

Postal Code     Borough                 Neighbourhood   Latitude  Longitude  \
8         M2J  North York  Fairview, Henry Farm, Oriole  43.778517 -79.346556   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
8               5        Clothing Store           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
8  Fast Food Restaurant            Restaurant             Juice Bar

# Report:

As there aren't many distinguishing factors, seems like many neighbourhoods in North York appear to be similar (assuming there was enough data collected). However, there are many factors that I feel makes this analysis inconclusive. Many of these neighbourhoods don't have many Venues. Had there been many more, the analysis would make a lot more sense, and perhaps had been more interesting. There may be a better Borough to analyze in the future.

Group 1: Quite possible that the magnitudes of both groups were small or near 0 (compared to the rest).
Group 2: Inconlusive, seems similar to group 4. Shopping centers?
Group 3: They share a bank. Not sure otherwise.
Group 4: Inconclusive, seems similar to group 2. Shopping centers.
Group 5: Many food shops.

Rather than comparing what's in the top group (lack of data), the separation between these clusters are perhaps more related to the magnitudes of the top five fractions. Additional study is needed.